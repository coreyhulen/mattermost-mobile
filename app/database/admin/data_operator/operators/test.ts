// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Q} from '@nozbe/watermelondb';

import {MM_TABLES} from '@constants/database';
import App from '@typings/database/app';

import DatabaseManager, {DatabaseType} from '../../database_manager';
import DataOperator from '../index';
import {IsolatedEntities, OperationType} from '../types';
import {
    operateAppRecord,
    operateCustomEmojiRecord,
    operateGlobalRecord,
    operateRoleRecord,
    operateServersRecord,
    operateSystemRecord,
    operateTermsOfServiceRecord,
} from './index';

jest.mock('../../database_manager');

const {APP} = MM_TABLES.DEFAULT;

describe('*** DataOperator: Operators tests ***', () => {
    const createConnection = async (setActive = false) => {
        const dbName = 'server_schema_connection';
        const serverUrl = 'https://appv2.mattermost.com';
        const database = await DatabaseManager.createDatabaseConnection({
            shouldAddToDefaultDatabase: true,
            databaseConnection: {
                actionsEnabled: true,
                dbName,
                dbType: DatabaseType.SERVER,
                serverUrl,
            },
        });

        if (setActive) {
            await DatabaseManager.setActiveServerDatabase({
                displayName: dbName,
                serverUrl,
            });
        }

        return database;
    };

    it('=> operateAppRecord: should return an array of type App ', async () => {
        expect.assertions(3);

        const database = await DatabaseManager.getDefaultDatabase();
        expect(database).toBeTruthy();

        const preparedRecords = await operateAppRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {
                buildNumber: 'build-7',
                createdAt: 1,
                id: 'id-18',
                versionNumber: 'v-1',
            },
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('App');
    });

    it('=> operateGlobalRecord: should return an array of type Global', async () => {
        expect.assertions(3);

        const database = await DatabaseManager.getDefaultDatabase();
        expect(database).toBeTruthy();

        const preparedRecords = await operateGlobalRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'g-1', name: 'g-n1', value: 'g-v1'},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('Global');
    });

    it('=> operateServersRecord: should return an array of type Servers', async () => {
        expect.assertions(3);

        const database = await DatabaseManager.getDefaultDatabase();
        expect(database).toBeTruthy();

        const preparedRecords = await operateServersRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {
                dbPath: 'mm-server',
                displayName: 's-displayName',
                id: 's-1',
                mentionCount: 1,
                unreadCount: 0,
                url: 'https://community.mattermost.com',
            },
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('Servers');
    });

    it('=> operateCustomEmojiRecord: should return an array of type CustomEmoji', async () => {
        expect.assertions(3);

        const database = await createConnection();
        expect(database).toBeTruthy();

        const preparedRecords = await operateCustomEmojiRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'emo-1', name: 'emoji'},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('CustomEmoji');
    });

    it('=> operateRoleRecord: should return an array of type Role ', async () => {
        expect.assertions(3);

        const database = await createConnection();
        expect(database).toBeTruthy();

        const preparedRecords = await operateRoleRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'role-1', name: 'role-name-1', permissions: []},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('Role');
    });

    it('=> operateSystemRecord: should return an array of type System', async () => {
        expect.assertions(3);

        const database = await createConnection();
        expect(database).toBeTruthy();

        const preparedRecords = await operateSystemRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'system-1', name: 'system-name-1', value: 'system'},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('System');
    });

    it('=> operateTermsOfServiceRecord: should return an array of type TermsOfService ', async () => {
        expect.assertions(3);

        const database = await createConnection();
        expect(database).toBeTruthy();

        const preparedRecords = await operateTermsOfServiceRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'system-1', acceptedAt: 1},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch(
            'TermsOfService',
        );
    });

    it('=> should create a record in the App table in the default database', async () => {
        expect.assertions(2);

        // Creates a record in the App table
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-1',
                    createdAt: 1,
                    id: 'id-1',
                    versionNumber: 'version-1',
                },
            ],
        });

        // Do a query and find out if the value has been registered in the App table of the default database
        const connection = await DatabaseManager.getDefaultDatabase();
        expect(connection).toBeTruthy();

        const records = (await connection!.collections.
            get(APP).
            query(Q.where('id', 'id-1')).
            fetch()) as App[];

        // We should expect to have a record returned as dictated by our query
        expect(records.length).toBe(1);
    });

    it('=> should create several records in the App table in the default database', async () => {
        expect.assertions(2);

        // Creates a record in the App table
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10',
                    createdAt: 1,
                    id: 'id-10',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11',
                    createdAt: 1,
                    id: 'id-11',
                    versionNumber: 'version-11',
                },
                {
                    buildNumber: 'build-12',
                    createdAt: 1,
                    id: 'id-12',
                    versionNumber: 'version-12',
                },
                {
                    buildNumber: 'build-13',
                    createdAt: 1,
                    id: 'id-13',
                    versionNumber: 'version-13',
                },
            ],
        });

        // Do a query and find out if the value has been registered in the App table of the default database
        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', Q.oneOf(['id-10', 'id-11', 'id-12', 'id-13']))).
            fetch()) as App[];

        // We should expect to have 4 records created
        expect(records.length).toBe(4);
    });

    it('=> should update a record in the App table in the default database', async () => {
        expect.assertions(3);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        // Update record having id 'id-1'
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.UPDATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-13-13',
                    createdAt: 1,
                    id: 'id-1',
                    versionNumber: 'version-1',
                },
            ],
        });

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', 'id-1')).
            fetch()) as App[];
        expect(records.length).toBeGreaterThan(0);

        // Verify if the buildNumber for this record has been updated
        expect(records[0].buildNumber).toMatch('build-13-13');
    });

    it('=> should update several records in the App table in the default database', async () => {
        expect.assertions(4);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        // Update records having id 'id-10' and 'id-11'
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.UPDATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10x',
                    createdAt: 1,
                    id: 'id-10',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11y',
                    createdAt: 1,
                    id: 'id-11',
                    versionNumber: 'version-11',
                },
            ],
        });

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', Q.oneOf(['id-10', 'id-11']))).
            fetch()) as App[];
        expect(records.length).toBe(2);

        // Verify if the buildNumber for those two record has been updated
        expect(records[0].buildNumber).toMatch('build-10x');
        expect(records[1].buildNumber).toMatch('build-11y');
    });

    it('=> [EDGE CASE] should UPDATE instead of CREATE record for existing id', async () => {
        expect.assertions(3);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        // id-10 and id-11 exist but yet the optType is CREATE.  The operator should then prepareUpdate the records instead of prepareCreate
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10x',
                    createdAt: 1,
                    id: 'id-10',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11x',
                    createdAt: 1,
                    id: 'id-11',
                    versionNumber: 'version-11',
                },
            ],
        });

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', Q.oneOf(['id-10', 'id-11']))).
            fetch()) as App[];

        // Verify if the buildNumber for those two record has been updated
        expect(records[0].buildNumber).toMatch('build-10x');
        expect(records[1].buildNumber).toMatch('build-11x');
    });

    it('=> [EDGE CASE] should CREATE instead of UPDATE record for non-existing id', async () => {
        expect.assertions(3);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        // id-15 and id-16 do not exist but yet the optType is UPDATE.  The operator should then prepareCreate the records instead of prepareUpdate
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.UPDATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10x',
                    createdAt: 1,
                    id: 'id-15',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11x',
                    createdAt: 1,
                    id: 'id-16',
                    versionNumber: 'version-11',
                },
            ],
        });

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', Q.oneOf(['id-15', 'id-16']))).
            fetch()) as App[];

        // Verify if the buildNumber for those two record has been created
        expect(records[0].buildNumber).toMatch('build-10x');
        expect(records[1].buildNumber).toMatch('build-11x');
    });

    // TODO :  new operators
});
