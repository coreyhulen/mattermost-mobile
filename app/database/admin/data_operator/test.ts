// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {MM_TABLES} from '@constants/database';

import DatabaseManager, {DatabaseType} from '../database_manager';
import DataOperator from './index';
import {
    operateAppRecord,
    operateCustomEmojiRecord,
    operateDraftRecord,
    operateGlobalRecord,
    operateRoleRecord,
    operateServersRecord,
    operateSystemRecord,
    operateTermsOfServiceRecord,
} from './operators';
import {OperationType, IsolatedEntities} from './types';

jest.mock('../database_manager');

const {DRAFT} = MM_TABLES.SERVER;

describe('*** DataOperator: Handlers tests ***', () => {
    const createConnection = async (setActive = false) => {
        const dbName = 'server_schema_connection';
        const serverUrl = 'https://appv2.mattermost.com';
        const database = await DatabaseManager.createDatabaseConnection({
            shouldAddToDefaultDatabase: true,
            databaseConnection: {
                actionsEnabled: true,
                dbName,
                dbType: DatabaseType.SERVER,
                serverUrl,
            },
        });

        if (setActive) {
            await DatabaseManager.setActiveServerDatabase({
                displayName: dbName,
                serverUrl,
            });
        }

        return database;
    };

    it('=> HandleApp: should write to APP entity', async () => {
        expect.assertions(2);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10x',
                    createdAt: 1,
                    id: 'id-21',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11y',
                    createdAt: 1,
                    id: 'id-22',
                    versionNumber: 'version-11',
                },
            ],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateAppRecord,
        });
    });

    it('=> HandleGlobal: should write to GLOBAL entity', async () => {
        expect.assertions(2);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.GLOBAL,
            values: [
                {id: 'global-1-id', name: 'global-1-name', value: 'global-1-value'},
            ],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateGlobalRecord,
        });
    });

    it('=> HandleServers: should write to SERVERS entity', async () => {
        expect.assertions(2);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.SERVERS,
            values: [
                {
                    dbPath: 'server.db',
                    displayName: 'community',
                    id: 'server-id-1',
                    mentionCount: 0,
                    unreadCount: 0,
                    url: 'https://community.mattermost.com',
                },
            ],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateServersRecord,
        });
    });

    it('=> HandleCustomEmoji: should write to CUSTOM_EMOJI entity', async () => {
        expect.assertions(2);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.CUSTOM_EMOJI,
            values: [
                {
                    id: 'custom-emoji-id-1',
                    name: 'custom-emoji-1',
                },
            ],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateCustomEmojiRecord,
        });
    });

    it('=> HandleRole: should write to ROLE entity', async () => {
        expect.assertions(2);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.ROLE,
            values: [
                {
                    id: 'custom-emoji-id-1',
                    name: 'custom-emoji-1',
                    permissions: ['custom-emoji-1'],
                },
            ],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateRoleRecord,
        });
    });

    it('=> HandleSystem: should write to SYSTEM entity', async () => {
        expect.assertions(2);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.SYSTEM,
            values: [{id: 'system-id-1', name: 'system-1', value: 'system-1'}],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateSystemRecord,
        });
    });

    it('=> HandleTermsOfService: should write to TERMS_OF_SERVICE entity', async () => {
        expect.assertions(2);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.TERMS_OF_SERVICE,
            values: [{id: 'tos-1', acceptedAt: 1}],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateTermsOfServiceRecord,
        });
    });

    it('=> No table name: should not call handleBase if tableName is invalid', async () => {
        expect.assertions(2);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: 'INVALID_TABLE_NAME',
            values: [{id: 'tos-1', acceptedAt: 1}],
        };

        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledTimes(0);
    });

    it('=> handleReactions: should write to both Reactions and CustomEmoji entities', async () => {
        expect.assertions(3);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnPrepareBase = jest.spyOn(DataOperator as any, 'prepareBase');
        const spyOnBatchOperation = jest.spyOn(DataOperator as any, 'batchOperations');

        await DataOperator.handleReactions({
            reactions: [
                {
                    create_at: 1608263728086,
                    delete_at: 0,
                    emoji_name: 'p4p1',
                    post_id: '4r9jmr7eqt8dxq3f9woypzurry',
                    update_at: 1608263728077,
                    user_id: 'ooumoqgq3bfiijzwbn8badznwc',
                },
            ],
            prepareRowsOnly: false,
        });

        // Called twice:  Once for Reaction record and once for CustomEmoji record
        expect(spyOnPrepareBase).toHaveBeenCalledTimes(2);

        // Only one batch operation for both entities
        expect(spyOnBatchOperation).toHaveBeenCalledTimes(1);
    });

    it('=> handleDraft: should write to the Draft entity', async () => {
        expect.assertions(2);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');
        const data = [
            {
                channel_id: '4r9jmr7eqt8dxq3f9woypzurrychannelid',
                files: [{
                    user_id: 'user_id',
                    post_id: 'post_id',
                    create_at: 123,
                    update_at: 456,
                    delete_at: 789,
                    name: 'an_image',
                    extension: 'jpg',
                    size: 10,
                    mime_type: 'image',
                    width: 10,
                    height: 10,
                    has_preview_image: false,
                    clientId: 'clientId',
                }],
                message: 'test draft message for post',
                root_id: '',
            },
        ];
        await DataOperator.handleDraft(data);

        // Only one batch operation for both entities
        expect(spyOnHandleBase).toHaveBeenCalledWith({
            optType: OperationType.CREATE,
            tableName: DRAFT,
            values: data,
            recordOperator: operateDraftRecord,
        });
    });

    it('=> handleFiles: should write to File entity', async () => {
        expect.assertions(3);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnPrepareBase = jest.spyOn(DataOperator as any, 'prepareBase');
        const spyOnBatchOperation = jest.spyOn(DataOperator as any, 'batchOperations');

        await DataOperator.handleFiles({
            files: [{
                user_id: 'user_id',
                post_id: 'post_id',
                create_at: 12345,
                update_at: 456,
                delete_at: 789,
                name: 'an_image',
                extension: 'jpg',
                size: 10,
                mime_type: 'image',
                width: 10,
                height: 10,
                has_preview_image: false,
            }],
            prepareRowsOnly: false,
        });

        expect(spyOnPrepareBase).toHaveBeenCalledTimes(1);
        expect(spyOnBatchOperation).toHaveBeenCalledTimes(1);
    });

    it('=> handlePostMetadata: should write to PostMetadata entity', async () => {
        expect.assertions(3);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnPrepareBase = jest.spyOn(DataOperator as any, 'prepareBase');
        const spyOnBatchOperation = jest.spyOn(DataOperator as any, 'batchOperations');

        const data = {
            images: {
                'https://community-release.mattermost.com/api/v4/image?url=https%3A%2F%2Favatars1.githubusercontent.com%2Fu%2F6913320%3Fs%3D400%26v%3D4': {
                    width: 400,
                    height: 400,
                    format: 'png',
                    frame_count: 0,
                },

            },
            embeds: [
                {
                    type: 'opengraph',
                    url: 'https://github.com/mickmister/mattermost-plugin-default-theme',
                    data: {
                        type: 'object',
                        url: 'https://github.com/mickmister/mattermost-plugin-default-theme',
                        title: 'mickmister/mattermost-plugin-default-theme',
                        description: 'Contribute to mickmister/mattermost-plugin-default-theme development by creating an account on GitHub.',
                        determiner: '',
                        site_name: 'GitHub',
                        locale: '',
                        locales_alternate: null,
                        images: [
                            {
                                url: '',
                                secure_url: 'https://community-release.mattermost.com/api/v4/image?url=https%3A%2F%2Favatars1.githubusercontent.com%2Fu%2F6913320%3Fs%3D400%26v%3D4',
                                type: '',
                                width: 0,
                                height: 0,
                            },
                        ],
                        audios: null,
                        videos: null,
                    },
                },
            ],
        };

        await DataOperator.handlePostMetadata({
            embeds: [{
                embed: data.embeds, postId: 'post-1',
            }],
            images: [{
                images: data.images, postId: 'post-1',
            }],
            prepareRowsOnly: false});

        expect(spyOnPrepareBase).toHaveBeenCalledTimes(1);
        expect(spyOnBatchOperation).toHaveBeenCalledTimes(1);
    });

    // TODO : test utils functions (  sanitizeReactions, createPostsChain, sanitizePosts)
});
