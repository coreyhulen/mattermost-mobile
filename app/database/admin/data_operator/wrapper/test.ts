// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import DatabaseManager from '@database/admin/database_manager';
import {createDataOperator} from '@database/admin/data_operator/wrapper';
import DatabaseConnectionException from '@database/admin/exceptions/database_connection_exception';
import {DatabaseType} from '@typings/database/enums';

jest.mock('@database/admin/database_manager');

describe('*** DataOperator Wrapper ***', () => {
    it('=> wrapper should return an instance of DataOperator ', async () => {
        expect.assertions(1);

        const serverUrl = 'https://wrapper.mattermost.com';

        // first we create the connection and save it into default database
        await DatabaseManager.createDatabaseConnection({
            configs: {
                actionsEnabled: true,
                dbName: 'community mattermost',
                dbType: DatabaseType.SERVER,
                serverUrl,
            },
            shouldAddToDefaultDatabase: true,
        });

        const dataOperator = await createDataOperator(serverUrl);

        expect(dataOperator).toBeTruthy();
    });

    it('=> wrapper should throw an error due to invalid server url', async () => {
        expect.assertions(2);

        const serverUrl = 'https://wrapper.mattermost.com';

        // first we create the connection and save it into default database
        await DatabaseManager.createDatabaseConnection({
            configs: {
                actionsEnabled: true,
                dbName: 'test_database',
                dbType: DatabaseType.SERVER,
                serverUrl,
            },
            shouldAddToDefaultDatabase: true,
        });

        await expect(createDataOperator('https://error.com')).rejects.toThrow(
            'No database has been registered with this url: https://error.com',
        );

        await expect(createDataOperator('https://error.com')).rejects.toThrow(DatabaseConnectionException);
    });
});
