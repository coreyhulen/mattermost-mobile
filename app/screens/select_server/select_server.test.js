// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {Appearance} from 'react-native-appearance';

import {shallowWithIntl} from 'test/intl-test-helper';

import {darkColors, lightColors} from 'app/styles/colors';
import SelectServer from './select_server';

jest.mock('react-native-splash-screen', () => ({
    hide: jest.fn(),
}));

describe('SelectServer', () => {
    const baseProps = {
        actions: {
            getPing: jest.fn(),
            handleServerUrlChanged: jest.fn(),
            handleSuccessfulLogin: jest.fn(),
            scheduleExpiredNotification: jest.fn(),
            loadConfigAndLicense: jest.fn(),
            login: jest.fn(),
            resetPing: jest.fn(),
            setLastUpgradeCheck: jest.fn(),
            setServerVersion: jest.fn(),
        },
        hasConfigAndLicense: false,
        serverUrl: '',
    };

    test('should show light background when user has light color scheme set', () => {
        Appearance.getColorScheme.mockImplementation(() => 'light');
        Appearance.addChangeListener.mockImplementation(() => 'light');

        const wrapper = shallowWithIntl(<SelectServer {...baseProps}/>);

        expect(wrapper.getElement()).toMatchSnapshot();
        expect(wrapper).toHaveStyle('backgroundColor', lightColors.containerBg);
    });

    test('should show dark background when user has dark color scheme set', () => {
        Appearance.getColorScheme.mockImplementation(() => 'dark');
        Appearance.addChangeListener.mockImplementation(() => 'dark');

        const wrapper = shallowWithIntl(<SelectServer {...baseProps}/>);

        expect(wrapper.getElement()).toMatchSnapshot();
        expect(wrapper).toHaveStyle('backgroundColor', darkColors.containerBg);
    });
});
