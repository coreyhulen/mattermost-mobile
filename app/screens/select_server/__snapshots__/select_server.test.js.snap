// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectServer should show dark background when user has dark color scheme set 1`] = `
<ForwardRef(_SafeAreaView)
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "#171717",
      },
    ]
  }
>
  <KeyboardAvoidingView
    behavior="padding"
    enabled={true}
    keyboardVerticalOffset={0}
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "backgroundColor": "#171717",
        },
      ]
    }
  >
    <StatusBar
      animated={false}
      barStyle="dark-content"
      showHideTransition="fade"
    />
    <TouchableWithoutFeedback
      onPress={[Function]}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "center",
            },
            Object {
              "paddingLeft": 15,
              "paddingRight": 15,
            },
            Object {
              "backgroundColor": "#171717",
            },
          ]
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../dist/assets/images/logo_dark.png",
            }
          }
        />
        <View>
          <FormattedText
            defaultMessage="Enter Server URL"
            id="mobile.components.select_server_view.enterServerUrl"
            style={
              Array [
                Object {
                  "fontSize": 32,
                  "fontWeight": "600",
                  "marginBottom": 15,
                  "marginTop": 15,
                  "textAlign": "center",
                },
                Object {
                  "fontSize": 20,
                  "fontWeight": "400",
                },
                Object {
                  "color": "#dbdbdb",
                },
              ]
            }
          />
        </View>
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          autoCorrect={false}
          disableFullscreenUI={true}
          editable={false}
          keyboardType="url"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onSubmitEditing={[Function]}
          placeholder="https://mattermost.example.com"
          placeholderTextColor="#9b9b9b"
          rejectResponderTermination={true}
          returnKeyType="go"
          style={
            Array [
              Object {
                "alignSelf": "stretch",
                "borderRadius": 3,
                "borderWidth": 2,
                "color": "#3d3c40",
                "fontSize": 16,
                "height": 45,
                "marginBottom": 5,
                "marginTop": 5,
                "paddingLeft": 10,
              },
              Object {
                "borderColor": "#9b9b9b",
                "color": "#dbdbdb",
              },
              Object {
                "borderWidth": 0,
              },
              Object {
                "backgroundColor": "#2e2e2e",
                "color": "#9b9b9b",
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <Connect(ErrorText)
          error={null}
        />
        <Button
          containerStyle={
            Array [
              Object {
                "alignItems": "center",
                "alignSelf": "stretch",
                "borderRadius": 3,
                "marginTop": 10,
                "padding": 15,
              },
              Object {
                "backgroundColor": "#399fff",
              },
              Object {
                "backgroundColor": "#2e2e2e",
              },
            ]
          }
          disabled={true}
          onPress={[Function]}
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 17,
                  "fontWeight": "500",
                  "textAlign": "center",
                },
                Object {
                  "color": "#171717",
                },
                Object {
                  "color": "#9b9b9b",
                },
              ]
            }
          >
            <FormattedText
              defaultMessage="Connect"
              id="mobile.components.select_server_view.connect"
            />
          </Text>
        </Button>
      </View>
    </TouchableWithoutFeedback>
  </KeyboardAvoidingView>
</ForwardRef(_SafeAreaView)>
`;

exports[`SelectServer should show light background when user has light color scheme set 1`] = `
<ForwardRef(_SafeAreaView)
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "#ffffff",
      },
    ]
  }
>
  <KeyboardAvoidingView
    behavior="padding"
    enabled={true}
    keyboardVerticalOffset={0}
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "backgroundColor": "#ffffff",
        },
      ]
    }
  >
    <StatusBar
      animated={false}
      barStyle="dark-content"
      showHideTransition="fade"
    />
    <TouchableWithoutFeedback
      onPress={[Function]}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "center",
            },
            Object {
              "paddingLeft": 15,
              "paddingRight": 15,
            },
            Object {
              "backgroundColor": "#ffffff",
            },
          ]
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../dist/assets/images/logo_light.png",
            }
          }
        />
        <View>
          <FormattedText
            defaultMessage="Enter Server URL"
            id="mobile.components.select_server_view.enterServerUrl"
            style={
              Array [
                Object {
                  "fontSize": 32,
                  "fontWeight": "600",
                  "marginBottom": 15,
                  "marginTop": 15,
                  "textAlign": "center",
                },
                Object {
                  "fontSize": 20,
                  "fontWeight": "400",
                },
                Object {
                  "color": "#171717",
                },
              ]
            }
          />
        </View>
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          autoCorrect={false}
          disableFullscreenUI={true}
          editable={false}
          keyboardType="url"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onSubmitEditing={[Function]}
          placeholder="https://mattermost.example.com"
          placeholderTextColor="#5c5c5c"
          rejectResponderTermination={true}
          returnKeyType="go"
          style={
            Array [
              Object {
                "alignSelf": "stretch",
                "borderRadius": 3,
                "borderWidth": 2,
                "color": "#3d3c40",
                "fontSize": 16,
                "height": 45,
                "marginBottom": 5,
                "marginTop": 5,
                "paddingLeft": 10,
              },
              Object {
                "borderColor": "#dbdbdb",
                "color": "#171717",
              },
              Object {
                "borderWidth": 0,
              },
              Object {
                "backgroundColor": "#dbdbdb",
                "color": "#5c5c5c",
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <Connect(ErrorText)
          error={null}
        />
        <Button
          containerStyle={
            Array [
              Object {
                "alignItems": "center",
                "alignSelf": "stretch",
                "borderRadius": 3,
                "marginTop": 10,
                "padding": 15,
              },
              Object {
                "backgroundColor": "#1247b1",
              },
              Object {
                "backgroundColor": "#dbdbdb",
              },
            ]
          }
          disabled={true}
          onPress={[Function]}
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 17,
                  "fontWeight": "500",
                  "textAlign": "center",
                },
                Object {
                  "color": "#ffffff",
                },
                Object {
                  "color": "#5c5c5c",
                },
              ]
            }
          >
            <FormattedText
              defaultMessage="Connect"
              id="mobile.components.select_server_view.connect"
            />
          </Text>
        </Button>
      </View>
    </TouchableWithoutFeedback>
  </KeyboardAvoidingView>
</ForwardRef(_SafeAreaView)>
`;
